File: client-py/test_protocol.py
"""Testes do parser e do encode: python3 -m unittest discover -v"""

import unittest

from protocol import Refs, encode_action, parse_cell, parse_line, parse_observation


class TestParseObservation(unittest.TestCase):
    """E1: uma linha OBS crua vira observação de domínio, resultado tagueado."""

    def test_e1_linha_obs_valida_vira_observacao_de_dominio(self):
        tag, obs = parse_observation("OBS srv-97 97 alive 97 9,5,plain,0,0;10,5,plain,3,1")
        self.assertEqual(tag, "ok")
        self.assertEqual(obs["tick"], 97)
        self.assertEqual(obs["status"], "alive")
        self.assertEqual(obs["scores_tick"], 97)
        self.assertEqual(len(obs["cells"]), 2)
        self.assertEqual(obs["cells"][1], {
            "x": 10, "y": 5, "terrain": "plain", "owner": 3, "fortified": 1,
        })

    def test_e1_token_extra_no_final_e_ignorado(self):
        tag, obs = parse_observation("OBS srv-1 1 alive 1 1,2,plain,0,0 lixo")
        self.assertEqual(tag, "ok")
        self.assertEqual(len(obs["cells"]), 1)

    def test_e1_obs_sem_tick_e_erro(self):
        tag, _ = parse_observation("OBS srv-1 alive 1")
        self.assertEqual(tag, "error")

    def test_e1_tag_diferente_de_obs_e_erro(self):
        tag, _ = parse_observation("WELCOME srv-0 3 aurora 96CDFB 60 40 1000 3 4,34")
        self.assertEqual(tag, "error")

    def test_e1_celula_forest_com_dono_expoe_terreno_e_owner(self):
        c = parse_cell("11,5,forest,2,0")
        self.assertEqual(c["terrain"], "forest")
        self.assertEqual(c["owner"], 2)


class TestEncode(unittest.TestCase):
    """E2: codificação das linhas do cliente."""

    def test_e2_encode_action_cobre_os_quatro_kinds(self):
        self.assertEqual(
            encode_action({"kind": "expand", "x": 1, "y": 2}, "r-1"),
            "ACT r-1 expand 1 2",
        )
        self.assertEqual(
            encode_action({"kind": "attack", "x": 3, "y": 4}, "r-2"),
            "ACT r-2 attack 3 4",
        )
        self.assertEqual(
            encode_action({"kind": "fortify", "x": 5, "y": 6}, "r-3"),
            "ACT r-3 fortify 5 6",
        )
        self.assertEqual(encode_action({"kind": "pass"}, "r-4"), "ACT r-4 pass")

    def test_e2_pass_omite_coordenadas(self):
        line = encode_action({"kind": "pass"}, "r-9")
        self.assertNotIn(",", line)
        self.assertEqual(len(line.split(" ")), 3)

    def test_e2_refs_mantem_o_segmento_de_sessao_e_incrementam_n(self):
        refs = Refs("aurora", session="k3f9")
        self.assertEqual(refs.next(), "aurora-k3f9-1")
        self.assertEqual(refs.next(), "aurora-k3f9-2")
        self.assertEqual(refs.session, "k3f9")


class TestParseLine(unittest.TestCase):
    """demais linhas do servidor e o leitor tolerante"""

    def test_welcome_de_colonia(self):
        tag, msg = parse_line("WELCOME srv-0 3 aurora 96CDFB 60 40 1000 3 4,34")
        self.assertEqual(tag, "ok")
        self.assertEqual(msg["type"], "welcome")
        self.assertEqual(msg["role"], "colony")
        self.assertEqual(msg["id"], 3)
        self.assertEqual(msg["name"], "aurora")
        self.assertEqual(msg["color"], "96CDFB")
        self.assertEqual(
            (msg["w"], msg["h"], msg["tick_ms"], msg["view_radius"]),
            (60, 40, 1000, 3),
        )
        self.assertEqual(msg["spawn"], [4, 34])

    def test_obs_via_parse_line(self):
        tag, msg = parse_line("OBS srv-97 97 alive 97 9,5,plain,0,0;10,5,plain,3,1")
        self.assertEqual(tag, "ok")
        self.assertEqual(msg["tick"], 97)
        self.assertEqual(msg["status"], "alive")
        self.assertEqual(msg["scores_tick"], 97)
        self.assertEqual(msg["cells"], [
            {"x": 9, "y": 5, "terrain": "plain", "owner": 0, "fortified": 0},
            {"x": 10, "y": 5, "terrain": "plain", "owner": 3, "fortified": 1},
        ])

    def test_ack_com_tokens_extra_ignorados(self):
        tag, msg = parse_line("ACK aurora-k3f9-17 97 extra tokens")
        self.assertEqual(tag, "ok")
        self.assertEqual(msg, {"type": "ack", "ref": "aurora-k3f9-17", "tick": 97})

    def test_nack(self):
        tag, msg = parse_line("NACK aurora-k3f9-17 too_late tick 96 resolvido")
        self.assertEqual(tag, "ok")
        self.assertEqual(msg["code"], "too_late")
        self.assertEqual(msg["detail"], "tick 96 resolvido")

    def test_score(self):
        tag, msg = parse_line("SCORE srv-97 97 3,aurora,21,alive;5,nova,14,dead")
        self.assertEqual(tag, "ok")
        self.assertEqual(msg["entries"][1], {
            "id": 5, "name": "nova", "cells": 14, "status": "dead",
        })

    def test_linhas_malformadas_viram_erro_nunca_excecao(self):
        self.assertEqual(parse_line("ACTN foo bar")[0], "error")
        self.assertEqual(parse_line("OBS srv-97 x alive 97")[0], "error")
        self.assertEqual(
            parse_line("WELCOME srv-0 3 aurora 96CDFB 60 40 1000 3")[0], "error",
        )
        self.assertIsNone(parse_cell("9,5,plain,0"))

    def test_encode_com_refs(self):
        refs = Refs("aurora", session="k3f9")
        self.assertEqual(refs.next(), "aurora-k3f9-1")
        self.assertEqual(
            encode_action({"kind": "expand", "x": 12, "y": 7}, refs.next()),
            "ACT aurora-k3f9-2 expand 12 7",
        )
        self.assertEqual(
            encode_action({"kind": "pass"}, refs.next()), "ACT aurora-k3f9-3 pass",
        )


if __name__ == "__main__":
    unittest.main()

File: client-c/kit/test/test_protocol.c
/* Testes de E1 (parse_observation) e E2 (encode_action), mais os parses
 * prontos. Sem framework: CHECK conta e imprime, main devolve != 0 se falhar.
 *
 * Rode com `make test`.
 */

#include <stdio.h>
#include <string.h>

#include "../src/protocol.h"

static int checks = 0, failures = 0;

#define CHECK(cond)                                                            \
  do {                                                                         \
    checks++;                                                                  \
    if (!(cond)) {                                                             \
      failures++;                                                              \
      printf("FALHOU %s:%d: %s\n", __FILE__, __LINE__, #cond);                 \
    }                                                                          \
  } while (0)

static Msg msg; /* grande, fora da pilha */

static void test_welcome(void) {
  CHECK(parse_line("WELCOME srv-0 3 aurora 96CDFB 60 40 1000 3 4,34", &msg) == 0);
  CHECK(msg.type == MSG_WELCOME);
  CHECK(msg.id == 3 && strcmp(msg.name, "aurora") == 0);
  CHECK(strcmp(msg.color, "96CDFB") == 0);
  CHECK(msg.w == 60 && msg.h == 40 && msg.tick_ms == 1000 && msg.view_radius == 3);
  CHECK(msg.spawn_x == 4 && msg.spawn_y == 34);
}

/* E1: estes testes ficam verdes quando parse_observation estiver pronta */
static void test_observation(void) {
  CHECK(parse_line("OBS srv-97 97 alive 97 9,5,plain,0,0;10,5,plain,3,1", &msg) == 0);
  CHECK(msg.type == MSG_OBS);
  CHECK(msg.tick == 97 && strcmp(msg.status, "alive") == 0 && msg.scores_tick == 97);
  CHECK(msg.n_cells == 2);
  if (msg.n_cells == 2) {
    CHECK(msg.cells[0].x == 9 && msg.cells[0].y == 5 && msg.cells[0].owner == 0);
    CHECK(msg.cells[1].owner == 3 && msg.cells[1].fortified == 1);
    CHECK(strcmp(msg.cells[1].terrain, "plain") == 0);
  }

  /* token extra no final é ignorado (drill da v2) */
  CHECK(parse_line("OBS srv-97 97 alive 97 9,5,plain,0,0 campo_extra", &msg) == 0);

  /* célula com sexto campo (v2) também parseia */
  CHECK(parse_cell("9,5,plain,0,0,42", &msg.cells[0]) == 0);

  /* colônia morta, visão vazia */
  CHECK(parse_line("OBS srv-98 98 dead 97", &msg) == 0);
  CHECK(strcmp(msg.status, "dead") == 0 && msg.n_cells == 0);

  /* malformadas viram erro, nunca crash */
  CHECK(parse_line("OBS srv-97 x alive 97", &msg) != 0);
  CHECK(parse_line("OBS srv-97 97 undead 97", &msg) != 0);
  CHECK(parse_line("OBS srv-97 97 alive 97 9,5,plain,0", &msg) != 0);
  CHECK(parse_cell("9,5,plain,0", &msg.cells[0]) != 0);
}

static void test_tolerant_and_malformed(void) {
  CHECK(parse_line("ACK aurora-k3f9-17 97 extra tokens", &msg) == 0);
  CHECK(msg.type == MSG_ACK && msg.tick == 97);
  CHECK(strcmp(msg.ref, "aurora-k3f9-17") == 0);

  CHECK(parse_line("ACTN foo bar", &msg) != 0);
  CHECK(parse_line("WELCOME srv-0 3 aurora 96CDFB 60 40 1000 3", &msg) != 0);
  CHECK(parse_line("ACK ref notanumber", &msg) != 0);
}

static void test_nack_score(void) {
  CHECK(parse_line("NACK aurora-k3f9-17 too_late tick 96 resolvido", &msg) == 0);
  CHECK(msg.type == MSG_NACK && strcmp(msg.code, "too_late") == 0);
  CHECK(strcmp(msg.detail, "tick 96 resolvido") == 0);

  CHECK(parse_line("SCORE srv-97 97 3,aurora,21,alive;5,nova,14,dead", &msg) == 0);
  CHECK(msg.type == MSG_SCORE && msg.tick == 97 && msg.n_entries == 2);
  if (msg.n_entries == 2) {
    CHECK(msg.entries[1].id == 5 && msg.entries[1].cells == 14);
    CHECK(strcmp(msg.entries[1].name, "nova") == 0);
    CHECK(strcmp(msg.entries[1].status, "dead") == 0);
  }
}

/* E2: estes testes ficam verdes quando encode_action estiver pronta */
static void test_encode(void) {
  Refs refs;
  char ref[MAX_REF];
  char line[128];
  Action expand = {"expand", 12, 7};
  Action pass = {"pass", 0, 0};

  refs_init_session(&refs, "aurora", "k3f9");

  refs_next(&refs, ref);
  CHECK(strcmp(ref, "aurora-k3f9-1") == 0);
  encode_hello(line, sizeof line, ref, "aurora");
  CHECK(strcmp(line, "HELLO v1 aurora-k3f9-1 colony aurora") == 0);

  refs_next(&refs, ref);
  CHECK(strcmp(ref, "aurora-k3f9-2") == 0);
  encode_action(line, sizeof line, &expand, ref);
  CHECK(strcmp(line, "ACT aurora-k3f9-2 expand 12 7") == 0);

  /* pass omite as coordenadas */
  refs_next(&refs, ref);
  encode_action(line, sizeof line, &pass, ref);
  CHECK(strcmp(line, "ACT aurora-k3f9-3 pass") == 0);
}

int main(void) {
  test_welcome();
  test_observation();
  test_tolerant_and_malformed();
  test_nack_score();
  test_encode();
  printf("%d verificacoes, %d falhas\n", checks, failures);
  return failures ? 1 : 0;
}
